### PR TITLE
Add logic for Delta to Rate Conversion

### PR DESCRIPTION
**Pull request by hossain-rayhan**
**Description:** <Describe what has changed.>
Add the logic for converting Delta Sum to Rate (gauge).

**Link to tracking Issue:** #3751 

**Testing:** I am adding unit test to improve coverage. 

**Documentation:** README added

### CODE REVIEW

**Review by jmacd — APPROVED** [review]

**Comment by hossain-rayhan** [comment]
This is dependent on #4218

**Comment by bogdandrutu** [comment]
Please rebase

**Comment by hossain-rayhan** [comment]
Hi @bogdandrutu, I rebased this and added more unit tests to improve coverage. Would you please take a look? Thanks.

File: processor/deltatorateprocessor/config.go
Comment (mxiamxia) [inline processor/deltatorateprocessor/config.go:60]: Any specific reason that you define a `timeUnits` map and then convert it to `array`. Can you directly use define the array with the valid `StringTimeUnit`?

File: processor/deltatorateprocessor/processor.go
Comment (mxiamxia) [inline processor/deltatorateprocessor/processor.go:63]: Calculating rate also requires you to maintain the state. it looks you apply rate and only check the metric name are same which is not right. You should consider dimensions & resource values should all be matched as well.

File: processor/deltatorateprocessor/config.go
Comment (hossain-rayhan) [reply processor/deltatorateprocessor/config.go]: Thanks. Updated.

File: processor/deltatorateprocessor/testdata/config_invalid_time_unit.yaml
Comment (hossain-rayhan) [reply processor/deltatorateprocessor/testdata/config_invalid_time_unit.yaml:1]: Added. Thanks for the suggestion.

File: processor/deltatorateprocessor/config.go
Comment (hossain-rayhan) [reply processor/deltatorateprocessor/config.go:60]: Hi Min, thanks for the review. Map is faster to lookup. Thats why the default is a map. This array is only used for error cases to print the message properly. Actually, this was suggested by Anthony for my previous PR. I want to stay with this if you don't think its a major concern.

File: processor/deltatorateprocessor/processor.go
Comment (hossain-rayhan) [reply processor/deltatorateprocessor/processor.go:63]: Hi Min, I want to explain a little bit here. I am not sure if I missed anything. This processor only coverts Delta to Rate. Our product requirement is to convert Cumulative Sum to Rate. We divided this into two part. 

1. Cumulative to Delta [We need the state here]. We update the usage and the startTimeStamp.
1. Delta to Rate [We don't need the state here]. We alreay have the usage and the time between startTimeStamp and current timeStamp. We can get the rate following `(usage) / (timeStamp - startTimeStamp)`. This was reviewed by Josh and Bogdan. Please check the (design doc)[https://docs.google.com/document/d/1NjeqzamzooqQP26hy5BWPvNsONLvdmV1bWqGwhMFoS4/edit#heading=h.br6h8f8h77c6]

**Comment by bogdandrutu** [comment]
@hossain-rayhan needs a rebase and also address all comments (mark them as resolve when that is the case).

**Comment by hossain-rayhan** [comment]
> @hossain-rayhan needs a rebase and also address all comments (mark them as resolve when that is the case).

Waiting for reviewer's feedback. I will mention you when all of them are resolved and we are ready to go. Thanks.

File: processor/deltatorateprocessor/processor.go
Comment (mxiamxia) [reply processor/deltatorateprocessor/processor.go:63]: Got it. you directly calculate the rate against the selected metrics with delta value attached. Thanks.

**Review by mxiamxia — APPROVED** [review]

**Review by anuraaga — APPROVED** [review]

**Comment by hossain-rayhan** [comment]
Hi @bogdandrutu, I rebased and got all the approvals. I think we are `ready_to_merge`. Please have a look.

**Comment by hossain-rayhan** [comment]
Hi @bogdandrutu, I rebased this PR. Can we get this merged now. Thanks.

File: processor/deltatorateprocessor/config.go
Comment (bogdandrutu) [inline processor/deltatorateprocessor/config.go:32]: Is rate usually per second? Can we not have this for the moment?

File: processor/deltatorateprocessor/processor.go
Comment (bogdandrutu) [inline processor/deltatorateprocessor/processor.go:62]: Does this apply to all metrics? Do we need a way to specify which metrics to apply to?

File: processor/deltatorateprocessor/config.go
Comment (hossain-rayhan) [reply processor/deltatorateprocessor/config.go:32]: For CPU usage are recorded in nanoseconds and those are calculated over Nanoseconds in our default EKS Container Insights. But for other metrics like error count, network bytes are usually per second. To me it seems like good to have.

File: processor/deltatorateprocessor/processor.go
Comment (hossain-rayhan) [reply processor/deltatorateprocessor/processor.go:62]: We are getting the metric names from the config file and mathing them before applying the conversions. I am not sure if I answered your question. Can you explain a bit if I missed your point.

File: processor/deltatorateprocessor/config.go
Comment (bogdandrutu) [reply processor/deltatorateprocessor/config.go:32]: I am lost, can you explain? For CPU-usage that is reported as 300ns (delta), the rate will be 300/(end-starttime in seconds)

File: processor/deltatorateprocessor/processor.go
Comment (bogdandrutu) [reply processor/deltatorateprocessor/processor.go:62]: This is because the code is written opposite to the rest of the code. It should be:

```
if !ok {
  continue
}
...
```

This also removes one indentation.

Comment (hossain-rayhan) [reply processor/deltatorateprocessor/processor.go:62]: Oh I see. You are smart. Will update.

File: processor/deltatorateprocessor/config.go
Comment (bogdandrutu) [reply processor/deltatorateprocessor/config.go:32]: If you want to get CPU seconds/second then actually will be strange to do this in one op. So need to convert the metric to seconds first.

File: processor/deltatorateprocessor/processor.go
Comment (bogdandrutu) [reply processor/deltatorateprocessor/processor.go:62]: Also `metrics` member name is bad, does not suggest anything about "filtering" :)

File: processor/deltatorateprocessor/config.go
Comment (hossain-rayhan) [reply processor/deltatorateprocessor/config.go:32]: I agree. If we need to convert the metrics, we can do that in an earlier stage using the metrics_geneartion processor scaling operation. 

For the rate calculations, we are giving customers the option. They knows which metrics are being recorded in what units. Say
1. CPU metric unit is in ns and time units are in ns. This one should be usage_ns/ time_difference_ns
1. Error Count metric unit in count and time units are in ns. For this one, maybe customer wants to calculate it per second.

File: processor/deltatorateprocessor/processor.go
Comment (hossain-rayhan) [reply processor/deltatorateprocessor/processor.go]: Pushed an update.

File: processor/deltatorateprocessor/config.go
Comment (hossain-rayhan) [reply processor/deltatorateprocessor/config.go:32]: I am not sure if the above mentioned use cases are convincing enought to have this option. Plese, let me know your opinion.

Comment (bogdandrutu) [reply processor/deltatorateprocessor/config.go:32]: > CPU metric unit is in ns and time units are in ns. This one should be usage_ns/ time_difference_ns

I can see how this time unit can work if the metric also has a unit time but is a very special case that we should probably  not worry about it in this processor.

> Error Count metric unit in count and time units are in ns. For this one, maybe customer wants to calculate it per second.

Which time units are in `ns`?

File: processor/deltatorateprocessor/processor.go
Comment (bogdandrutu) [inline processor/deltatorateprocessor/processor.go:66]: Probably we should have an error logged if this condition is not happening:

```
if metric.DataType() != pdata.MetricDataTypeSum || metric.Sum().AggregationTemporality() != pdata.AggregationTemporalityDelta {
  logger.Error("Configured metric for rate calculation that is not a sum or delta") // something these lines
  continue
}

```

File: processor/deltatorateprocessor/config.go
Comment (hossain-rayhan) [reply processor/deltatorateprocessor/config.go:32]: > Which time units are in `ns`?

`dp.StartTimestamp` and `dp.Timestamp` both are in `ns`, right?

File: processor/deltatorateprocessor/processor.go
Comment (hossain-rayhan) [reply processor/deltatorateprocessor/processor.go:66]: Should this be an error log. I feel like we can print a Debug message. Any thought?

File: processor/deltatorateprocessor/config.go
Comment (hossain-rayhan) [reply processor/deltatorateprocessor/config.go:32]: So whats your final suggestion? We will only support the default rate calculation per/second?

File: processor/deltatorateprocessor/processor.go
Comment (bogdandrutu) [reply processor/deltatorateprocessor/processor.go:66]: Not strong feeling, but probably at least INFO

File: processor/deltatorateprocessor/config.go
Comment (bogdandrutu) [reply processor/deltatorateprocessor/config.go:32]: > dp.StartTimestamp and dp.Timestamp both are in ns, right? indeed and always are like that. So you always have to transform them to second.

Comment (hossain-rayhan) [reply processor/deltatorateprocessor/config.go:32]: Thanks. I will remove the config option for now and make the default rate calculation per/second.

Comment (bogdandrutu) [reply processor/deltatorateprocessor/config.go:32]: waiting for the change :)

Comment (hossain-rayhan) [reply processor/deltatorateprocessor/config.go]: Pushed an update. Thanks.

File: processor/deltatorateprocessor/processor.go
Comment (hossain-rayhan) [reply processor/deltatorateprocessor/processor.go]: Pushed an update. Thanks.

File: processor/deltatorateprocessor/processor.go
Comment (bogdandrutu) [inline processor/deltatorateprocessor/processor.go:65]: s/is not a sum or delta/is not a delta sum

File: processor/deltatorateprocessor/processor.go
Comment (bogdandrutu) [inline processor/deltatorateprocessor/processor.go:64]: Should this be monotonic as well?

/cc @jmacd

File: processor/deltatorateprocessor/processor.go
Comment (bogdandrutu) [inline processor/deltatorateprocessor/processor.go:76]: cheaper to calculate:

`duration = (fromDataPoint.Timestamp() - fromDataPoint.StartTimestamp()) * time.Nanoseconds`

File: processor/deltatorateprocessor/processor.go
Comment (jmacd) [reply processor/deltatorateprocessor/processor.go:64]: 🤷  The output is well-defined for monotonic and non-monotonic cases, implying the rate will or won't be non-negative.

File: processor/deltatorateprocessor/processor.go
Comment (jmacd) [reply processor/deltatorateprocessor/processor.go:76]: Agree!

**Review by jmacd — APPROVED** [review]

File: processor/deltatorateprocessor/processor.go
Comment (hossain-rayhan) [reply processor/deltatorateprocessor/processor.go:64]: I am not sure if I understood. Can you please help me. Do we want to check whether its monotonic or it doesn't matter? Thanks.

File: processor/deltatorateprocessor/processor.go
Comment (hossain-rayhan) [reply processor/deltatorateprocessor/processor.go]: Updated.

File: processor/deltatorateprocessor/processor.go
Comment (hossain-rayhan) [reply processor/deltatorateprocessor/processor.go]: Updated.

**Comment by hossain-rayhan** [comment]
Hi @bogdandrutu any final comment? Can you please have a quick look?

**Pull request merged**
